lib/alerts/medicine_article_monitor.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+# This class to create alerts when a WikiProject Medicine article
+# is assigned for a course that doesn't have the medical content
+# training module.
+# wiki_education mode only
+# Goal: to catch cases where a class that wasn't expected 
+# to work on medical topics tries to do so.
+# Issue #4693
+class MedicineArticleMonitor
+  WIKI_PROJECT_MEDICINE_CAT = 'Category:All WikiProject Medicine articles'
+  NUMBER_OF_ARTICLES_TO_CHECK = 5
+
+  def self.create_alerts_for_no_med_training_for_course
+    new.create_alerts_for_no_med_training_for_course
+  end
+
+  def initialize
+    @wiki = Wiki.find_by(language: 'en', project: 'wikipedia')
+  end
+
+  def create_alerts_for_no_med_training_for_course
+    last_assigned_med_articles.each do |assignment|
+      unless med_training_for_course?(assignment.course_id)
+        create_alert_for_no_med_training_for_course(assignment)
+      end
+    end
+  end
+
+  def create_alert_for_no_med_training_for_course(assignment)
+    return if alert_already_exists?(assignment)
+
+    alert = Alert.create!(type: 'NoMedTrainingForCourseAlert',
+                  article_id: assignment.article_id,
+                  course_id: assignment.course_id)
+    alert.email_content_expert
+  end
+
+  def alert_already_exists?(assignment)
+    Alert.exists?(type: 'NoMedTrainingForCourseAlert',
+                  article_id: assignment.article_id,
+                  course_id: assignment.course_id)
+  end
+
+  def last_assigned_med_articles
+    last_assigned_articles
+      .select { |assignment| med_article?(assignment.article_title) }
+  end
+
+  def last_assigned_articles
+    Assignment.last NUMBER_OF_ARTICLES_TO_CHECK
+  end
+
+  def med_training_for_course?(course_id)
+    if Course
+       .find(course_id)
+       .training_modules
+       .find { |tm| tm.slug == 'editing-medical-topics' }
+      true
+    else
+      false
+    end
+  end
+
+  # Query
+  # curl "https://en.wikipedia.org/w/api.php?action=query
+  # &prop=categories&titles=Talk:Appendectomy
+  # &cllimit=50&clcategories=Category:All_WikiProject_Medicine_articles
+  # &indexpageids=1&format=json"
+  # No need to normalize article title
+  # "Talk:Colonic polypectomy" eq Talk:Colonic_polypectomy
+  def med_article?(title)
+    api = WikiApi.new @wiki
+    query_params = {
+      action: 'query',
+      prop: 'categories',
+      clcategories: WIKI_PROJECT_MEDICINE_CAT,
+      titles: "Talk:#{title}",
+      cllimit: 50,
```

Comment:
Hmm... this approach will do a separate API query for every title being checked.

There are a few alternatives that I suspect could be faster.

You can add multiple titles to the same query, doing a search for all of the recent article titles at once. There might be cases where we search for many medical articles at once, though, so the limit of 50 might come into play. (It can probably be raised to 500, which would make it unlikely to be a problem, although still maybe possible.)

Alternatively, you could use the `Category` model to store and occasionally update the entire list of WikiProject Medicine articles. That would look something like `Category.find_or_create_by(source: 'category', name: 'All_WikiProject_Medicine_articles', wiki: <English Wikipedia>)`. This category could be updated (via `refresh_titles`) if it hasn't been updated within the last day. (It takes about 30 seconds to refresh a category that large, so we probably don't want to do it every update cycle.)

I think the `Category` approach is probably the most appropriate here.

Reply:
I was a bit proud of that method, and this query :)
But, I see your point, I tried the `Category` thing.
The problem is; how to update the category ? We cannot put it in the constant_update.
Can it have its own worker in schedule.yml ? Or some other place ?

Reply:
I think it can be done from this file, with something like `category.refresh_titles unless category.updated_at > 1.day.ago`. Updating that category is something that only this monitor cares about, so I think it would be unnecessarily complex to schedule that update separately.

Reply:
I have changed to your Category solution, with this refreshing strategy.